openedx/core/djangoapps/oauth_dispatch/docs/decisions/0013-add-grant-type-in-jwt-payload.rst

```diff
@@ -0,0 +1,27 @@
+13. Added grant type in JWT payload
+-----------------------------------------
+
+Status
+------
+
+Accepted
+
+Context
+-------
+
+Edx mobile apps are `migrating from Dot access token to JWT token`_ authentication.Therefore gradually adding JWT support in required APIS.
+To exchange JWT with session cookies we need to determine the grant type of the access token first.
+We are already doing this in ``/oauth2/login`` view for dot access token, but we don't have grant type in JWT.
+
+.. _migrating from Dot access token to JWT token: https://2u-internal.atlassian.net/browse/LEARNER-8481
+
+Decisions
+---------
+
+A new key ``grant_type`` will be added in payload while creating JWT. Key ``grant_type`` will be determined through dot access token for mobile flow.
+Since we create JWT from access token in mobile flow therefore its value will be ``password``. We can provide any value other than password and by default it will be an empty string.
```

Comment:
Actually, could you manually test the following locally and share what you see in the JWT:
1. Call to access_token using a password.
2. Call to access_token using client credentials.
3. What you see in the JWT payload cookie in the browser.

This ADR note will probably need to be updated accordingly.

Reply:
When I called access token I got following grant_types:
with password: "password"
with client credentials: "client-credentials"
In browser jwt cookie I got "password" stored as grant_type.

Reply:
BTW you can test these changes on [this sandbox](iap.sandbox.edx.org/).

Reply:
> At this time, all values for grant type may not yet be supported, and an empty string will be used by default to represent any grant type that is not password. However, additional grant type values may be implemented as-needed.

This line is not 100% true since we are supporting all the grant types supported by DOT token. and will only have empty string when no grant type is provided.

Reply:
1. Note: I see the following in New Relic, but this is from the [POST param](https://github.com/openedx/edx-platform/blob/fe6b666f5b0bea853ba627e7cff4adbab2904a0d/openedx/core/djangoapps/oauth_dispatch/views.py#L105), and not what is in the DB:
```
Oauth Grant Type | Response.Status | Transactions
<BLANK>  | 200 | 919178436
refresh_token | 200 | 2427823
<BLANK>  | 204 | 1792403
<BLANK>  | 201 | 1620052
<BLANK>  | 206 | 755693
client_credentials | 200 | 155905
authorization_code | 200 | 145699
password | 200 | 98034
<BLANK>  | 202 | 1233
```
2. I still don't understand what a blank indicates, and whether this would be `''` for our new `grant_type`? If authentication is successful, doesn't some grant type needed to have been checked?
3. Is there a way to ensure that the JWT will have the right `grant_type` in the newer mobile JWT flow before we switch to it? Maybe this is obvious, but the New Relic data has me concerned. We should work out a dashboard to watch for launch, and make sure we have all the metrics we need.
4. We should remove the last consequence, or update it to clarify when the empty string will be used. It's still unclear to me what it means.

Reply:
Currently we are testing these changes in a sandbox environment and flow is working fine. Maybe we can put some extra analytics in sandbox to look for grant type in database. what do you say?

Reply:
If sandbox testing is going well, that's great. You are testing with a refresh token as well?

Separately, for final rollout, the question will be what you'll want to compare to see if successful mobile traffic seems comparable before/after. We may have all the data we need. Otherwise, we may want to add some data. I'm happy to help put something together with you at some point. This isn't urgent. Are you planning on using some sort of toggle so you can switch back and forth, in case of issues?

Reply:
We will start full testing with all devices and platforms from next week. Currently we just tested flow and it was working fine.
Yes we have a toggle to move back to Bearer. 
This looks great and will help us analyze data. I'll surely need your help on this in future.

Reply:
Note: I'm considering this resolved. We can take up rollout questions elsewhere.